File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
         int n=nums.size();
        k=k%n;//normalizing k prevents unnecessary extra rotations.Rotate by k=11: same as k=11%5=1 → [5,1,2,3,4]

        vector<int>temp(k);
       

        for(int i=0;i<k;i++){
            temp[i]=nums[n-k+i];
        }
        for(int i=n-1;i>=k;i--){
            nums[i]=nums[i-k];
        }
        
        for(int i=0;i<k;i++){
            nums[i]=temp[i];
            
        }
        
    }
};

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        #first reverse the n element (7,6,5,4,3,2,1)
        #second reverse the fisrt k element(5,6,7,4,3,2,1)
        #third reverse the n-k element (5,6,7,1,2,3,4)

        n=len(nums)
        k=k%n
        def reverse(start: int, end:int) -> None:
            while(start<end):
                nums[start],nums[end]=nums[end],nums[start]
                start+=1
                end-=1
            
        reverse(0,n-1)
        reverse(0,k-1)
        reverse(k,n-1)